src/Authoring/WinRT.SourceGenerator/MergeReferencedActivationFactoriesGenerator.cs

```diff
@@ -0,0 +1,199 @@
+        });
+
+        // Get the fully qualified type names of all assembly exports types to merge
+        IncrementalValueProvider<EquatableArray<string>> assemblyExportsTypeNames =
+            context.CompilationProvider
+            .Combine(isGeneratorEnabled)
+            .Select(static (item, token) =>
+            {
+                // Immediately bail if the generator is disabled
+                if (!item.Right)
+                {
+                    return new EquatableArray<string>(ImmutableArray<string>.Empty);
+                }
+
+                ImmutableArray<string>.Builder builder = ImmutableArray.CreateBuilder<string>();
+
+                foreach (MetadataReference metadataReference in item.Left.References)
+                {
+                    token.ThrowIfCancellationRequested();
+
+                    if (item.Left.GetAssemblyOrModuleSymbol(metadataReference) is not IAssemblySymbol assemblySymbol)
+                    {
+                        continue;
+                    }
+
+                    token.ThrowIfCancellationRequested();
+
+                    // Add the type name if the assembly is a WinRT component
+                    if (TryGetDependentAssemblyExportsTypeName(
+                        assemblySymbol,
+                        item.Left,
+                        token,
+                        out string? name))
+                    {
+                        builder.Add(name);
+                    }
+                }
+
+                token.ThrowIfCancellationRequested();
+
+                return new EquatableArray<string>(builder.ToImmutable());
+            });
+
+        // Generate the chaining helper
+        context.RegisterImplementationSourceOutput(assemblyExportsTypeNames, static (context, assemblyExportsTypeNames) =>
+        {
+            if (assemblyExportsTypeNames.IsEmpty)
+            {
+                context.AddSource("ChainedExports.g.cs", """
+                    // <auto-generated/>
+                    #pragma warning disable
+
+                    namespace WinRT
+                    {
+                        using global::System;
+
+                        /// <inheritdoc cref="Module"/>
+                        partial class Module
+                        {
+                            /// <summary>
+                            /// Tries to retrieve the activation factory from all dependent WinRT components.
+                            /// </summary>
+                            /// <param name="fullyQualifiedTypeName">The marshalled fully qualified type name of the activation factory to retrieve.</param>
+                            /// <returns>The pointer to the activation factory that corresponds with the class specified by <paramref name="fullyQualifiedTypeName"/>.</returns>
+                            internal static IntPtr TryGetDependentActivationFactory(ReadOnlySpan<char> fullyQualifiedTypeName)
```

Comment:
I feel like the code in the else would of just handled this scenario just fine as the loop just wouldn't run.  Or did I miss any other differences between both?

Reply:
I think I added this as an optimization in the default case, so we don't need to create the intermediate builder. But then again in the default case, this output node will only ever run once, so it's likely not needed. Removed!